components/Carousel/Carousel.jsx

```diff
@@ -0,0 +1,188 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import Icon from '../Icon';
+import { If, getAspectRatioHeight } from '../util';
+
+
+// calcNext(3, 0) => 1
+// calcNext(3, 1) => 2
+// calcNext(3, 2) => 0 // <- it wraps around to the first slide
+function calcNext(length, current) {
+  return (current + 1) % length;
+}
+
+// calcPrev(3, 2) => 1
+// calcPrev(3, 1) => 0
+// calcPrev(3, 0) => 2 // <- it wraps around to the last slide
+function calcPrev(length, current) {
+  return (current - 1) < 0 ? (length - 1) : (current - 1);
+}
+
+function containValue(max, min, value) {
+  if (value > max) { return max; }
+  if (value < min) { return min; }
+  return value;
+}
+
+
+class Carousel extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      bgSlideIndex: 1,
+      enterDirection: 'TO_LEFT',
+      exitDirection: '',
+      isFresh: true, // `isFresh` just means no slide change has been triggered yet. it's a hack used to allow a custom `enter` value on the first bgSlide. would like to find a better alternative to this...
+      topSlideIndex: 0,
+      isMobile: false, // passed down to <Slide>
+      height: 0, // passed down to <Slide> so it can reuse the h/w calculations
+      width: 0, // passed down to <Slide> so it can reuse the h/w calculations
+    };
+    this.el = null;
+    this.setProportionalHeight = this.setProportionalHeight.bind(this);
+    this.keyboardNavigate = this.keyboardNavigate.bind(this);
+    this.generateCoin = this.generateCoin.bind(this);
+    this.goToSlide = this.goToSlide.bind(this);
+    this.next = this.next.bind(this);
+    this.prev = this.prev.bind(this);
+  }
+
+  componentDidMount() {
+    this.setProportionalHeight();
+    window.addEventListener('resize', this.setProportionalHeight);
+    // NOTE: if keyboard navigation ends up being an issue because of <input> elements on the page,
+    // maybe bind the event to `this.el` instead of `window`.
+    window.addEventListener('keyup', this.keyboardNavigate);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.setProportionalHeight);
+    window.removeEventListener('keyup', this.keyboardNavigate);
+  }
+
+  setProportionalHeight() {
+    if (this.el) {
+      const MOBILE_PADDING_BOTTOM = 0;
+      const { maxHeight, minHeight } = this.props;
```

Comment:
Should we also add `aspectRatio` to this, instead of using `this.props.aspectRatio` on line 68?